README.md

# lit

<!-- badges: start -->

[![R-CMD-check](https://github.com/ajbass/lit/actions/workflows/R-CMD-check.yml/badge.svg)](https://github.com/ajbass/lit/actions/workflows/R-CMD-check.yml)
[![CRAN_Status_Badge](https://www.r-pkg.org/badges/version/lit)](https://cran.r-project.org/package=lit)

<!-- badges: end -->

## Overview

The `lit` package implements a kernel-based multivariate testing
procedure, called Latent Interaction Testing (LIT), to test for latent
genetic interactions in genome-wide association studies. See our
manuscript for additional details:

> Bass AJ, Bian S, Wingo AP, Wingo TS, Culter DJ, Epstein MP.
> Identifying latent genetic interactions in genome-wide association
> studies using multiple traits. *Submitted*; 2023.

## Installation

This software is implemented in the `R` statistical programming
language. To install the release version, type the following in the `R`
terminal:

``` r
# release version
install.packages("lit")
```

The development version of `lit` can be installed using the following
code:

``` r
# install devtools
install.packages("devtools")
devtools::install_github("ajbass/lit")
```

The vignette can be viewed by typing:

``` r
browseVignettes(package = "lit")
```

If you run into issues with `gfortran` on Mac, see the answer
[here](https://stackoverflow.com/questions/69639782/installing-gfortran-on-macbook-with-apple-m1-chip-for-use-in-r)
for additional details.

## Quick start

We provide two ways to use the `lit` package. When the genotypes can be
loaded in R (small GWAS datasets), the `lit()` function can be used:

``` r
library(lit)
# set seed
set.seed(123)

# generate 10 SNPs for 10 individuals
X <- matrix(rbinom(10 * 10, size = 2, prob = 0.25), ncol = 10)

# generate 4 phenotypes for 10 individuals
Y <- matrix(rnorm(10 * 4), ncol = 4) 

# test for latent genetic interactions
out <- lit(Y, X)
head(out)
#>        wlit      ulit      alit
#> 1 0.2681410 0.3504852 0.3056363
#> 2 0.7773637 0.3504852 0.6044655
#> 3 0.4034423 0.3504852 0.3760632
#> 4 0.7874949 0.3504852 0.6157108
#> 5 0.8701189 0.3504852 0.7337565
#> 6 0.2352616 0.3504852 0.2847600
```

The output is a data frame of p-values where the rows are SNPs and the
columns are different implementations of LIT to test for latent genetic
interactions:

- `wlit` uses a linear kernel to measure pairwise similarity for the
  genotype and trait matrices
- `ulit` uses a projection kernel to measure pairwise similarity for the
  genotype and trait matrices
- `alit` combines the p-values of `wlit` and `ulit` using a Cauchy
  combination test to maximize the number of discoveries

For large GWAS datasets (e.g., biobank-sized), the `lit()` function is
not computationally feasible because the genotypes cannot be loaded in
`R`. Instead, the `lit_plink()` function can be applied directly to
plink files. To demonstrate how to use the function, we use the example
plink files from the `genio` package:

``` r
# load genio package
library(genio)

# path to plink files
file <- system.file("extdata", 'sample.bed', package = "genio", mustWork = TRUE)

# generate trait expression
Y <- matrix(rnorm(10 * 4), ncol = 4)

# apply lit to plink file
out <- lit_plink(Y, file = file, verbose = FALSE)
head(out)
#>   chr         id     pos alt ref       maf      wlit      ulit      alit
#> 1   1  rs3094315  752566   G   A 0.3888889 0.7908763 0.3422960 0.6150572
#> 2   1  rs7419119  842013   T   G 0.3888889 0.1552580 0.3422960 0.2194972
#> 3   1 rs13302957  891021   G   A 0.2500000 0.4088937 0.3325939 0.3687589
#> 4   1  rs6696609  903426   C   T 0.3125000 0.5857829 0.3325939 0.4519475
#> 5   1     rs8997  949654   A   G 0.4375000 0.6628300 0.3325939 0.4969663
#> 6   1  rs9442372 1018704   A   G 0.2500000 0.3192430 0.3325939 0.3258332
```

See `?lit` and `?lit_plink` for additional details and input arguments.

Note that a marginal testing procedure for latent genetic interactions
based on the squared residuals and cross products (Marginal (SQ/CP)) can
also be implemented using the `marginal` and `marginal_plink` functions:

``` r
# apply Marginal (SQ/CP) to loaded genotypes
out <- marginal(Y, X)

# apply Marginal (SQ/CP) to plink file
out <- marginal_plink(Y, file = file, verbose = FALSE)
```

<!---
The columns represent the cross products, e.g., `col1_col2` is the cross product of the first trait multiplied by the second trait (adjusting for the additive genetic effect)). See our preprint for additional details on this approach.

## To do

List of vario
NA's in phenotypes.

-->


NEWS.md
# lit (development version)


# lit 1.0.0

* Initial CRAN submission.


R/lit.R
#' Latent Interaction Testing
#'
#' @description
#' \code{lit} performs a kernel-based testing procedure, Latent Interaction Testing (LIT), using a set of traits and SNPs.
#' LIT tests whether the squared residuals (SQ) and cross products (CP) are statistically independent of the genotypes.
#' In particular, we construct a kernel matrix for the SQ/CP terms to measure the pairwise
#' similarity between individuals, and also construct an analogous one for the genotypes.
#' We then test whether these two matrices are independent.
#' Currently, we implement the linear and projection kernel functions to measure pairwise similarity between individuals.
#' We then combine the p-values of these implementations using a Cauchy combination test to maximize the number of discoveries.
#'
#' @param y matrix of traits (n observations by k traits)
#' @param x matrix of SNPs (n observations by m SNPs)
#' @param adjustment matrix of covariates to adjust traits
#' @param pop_struct matrix of PCs that captures population structure
#'
#' @return
#' A data frame of p-values where the columns are
#'  \itemize{
#'  \item{\code{wlit}: }{LIT using a linear kernel}
#'  \item{\code{ulit}: }{LIT using a projection kernel}
#'  \item{\code{alit}: }{Cauchy combination test of the above two LIT implementations.}
#' }
#'
#' @examples
#' # set seed
#' set.seed(123)
#'
#' # Generate SNPs and traits
#' X <- matrix(rbinom(10*2, size = 2, prob = 0.25), ncol = 2)
#' Y <- matrix(rnorm(10*4), ncol = 4)
#'
#' out <- lit(Y, X)
#'
#' @seealso \code{\link{lit_plink}}
#' @export
lit <- function(y,
                x,
                adjustment = NULL,
                pop_struct = NULL) {
  if (!(is.matrix(y) & is.matrix(x))) {
    stop("Inputs must be matrices.")
  }
  if (!(nrow(y) == nrow(x))) {
    stop("Observations are rows for each input.")
  }
  if (anyNA(y)) {
    stop("Current implementation requires no NAs in trait matrix. Need to adjust inputs accordingly ... feature will be added soon.")
  }

  if (is.null(pop_struct)) {
    h = matrix(1, nrow = nrow(x), ncol = 1)
  } else {
    if (!(is.matrix(pop_struct) & (nrow(y) == nrow(pop_struct)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
    h = cbind(1, pop_struct)
  }

  if (is.null(adjustment)) {
    y <- .quick_lm_cpp(Xs = h, Ys = y)
  } else {
    if (!(is.matrix(adjustment) & (nrow(y) == nrow(adjustment)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
     y <- .quick_lm_cpp(Xs = cbind(h, adjustment), Ys = y)
  }
  m <- ncol(x)
  p <- matrix(nrow = m, ncol = 3)
  ind <- !is.na(x)
  for (i in 1:m) {
    indtmp <- ind[,i] # remove NAs from genotypes
    xtmp <- x[indtmp, i, drop=F]
    ytmp <- y[indtmp,,drop = F]
    htmp <- h[indtmp,,drop = F]
    p[i,] <- .lit_cpp(Xs = xtmp, Ys = ytmp, Hs = htmp)
  }
  colnames(p) <-  c("wlit", "ulit", "alit")
  as.data.frame(p)
}

#' LIT correcting for dominance effects
#'
#' @description
#' Internal use for now
#'
#' @param y matrix of traits (n observations by k traits)
#' @param x matrix of SNPs (n observations by m SNPs)
#' @param adjustment matrix of covariates to adjust traits
#' @param pop_struct matrix of PCs that captures population structure
lit_h <- function(y,
                  x,
                  adjustment = NULL,
                  pop_struct = NULL) {
  if (!(is.matrix(y) & is.matrix(x))) {
    stop("Inputs must be matrices.")
  }
  if (!(nrow(y) == nrow(x))) {
    stop("Observations are rows for each input.")
  }
  if (anyNA(y)) {
    stop("Current implementation requires no NAs in trait matrix. Need to adjust inputs accordingly ... feature will be added soon.")
  }

  if (is.null(pop_struct)) {
    h = matrix(1, nrow = nrow(x), ncol = 1)
  } else {
    if (!(is.matrix(pop_struct) & (nrow(y) == nrow(pop_struct)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
    h = cbind(1, pop_struct)
  }

  if (is.null(adjustment)) {
      y <- .quick_lm_cpp(Xs = h, Ys = y)
  } else {
    if (!(is.matrix(adjustment) & (nrow(y) == nrow(adjustment)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
    y <- .quick_lm_cpp(Xs = cbind(h, adjustment), Ys = y)
  }
  m <- ncol(x)
  p <- matrix(nrow = m, ncol = 3)
  ind <- !is.na(x)
  for (i in 1:m) {
    indtmp <- ind[,i] # remove NAs from genotypes
    xtmp <- x[indtmp, i, drop=F]
    ytmp <- y[indtmp,,drop = F]
    htmp <- h[indtmp,,drop = F]
    maf <-   mean(xtmp) / 2
    if (maf > 0.5) xtmp = 2 - xtmp
    m1 <- as.numeric(xtmp == 1)
    m2 <- as.numeric(xtmp == 2)
    xdom <- model.matrix(~htmp + m1 + m2)
    ytmp <- residuals(lm(ytmp ~ xdom))
    p[i,] <- .lit_cpp(Xs = xtmp, Ys = ytmp, Hs = htmp)
  }
  colnames(p) <-  c("wlit", "ulit", "alit")
  as.data.frame(p)
}

#' Latent Interaction Testing
#'
#' \code{lit_plink} performs a kernel-based testing procedure, Latent Interaction Testing (LIT), using a set of traits and SNPs.
#' LIT tests whether the squared residuals (SQ) and cross products (CP) are statistically independent of the genotypes.
#' In particular, we construct a kernel matrix for the SQ/CP terms to measure the pairwise
#' similarity between individuals, and also construct an analogous one for the genotypes.
#' We then test whether these two matrices are independent.
#' Currently, we implement the linear and projection kernel functions to measure pairwise similarity between individuals.
#' We then combine the p-values of these implementations using a Cauchy combination test to maximize the number of discoveries.
#' This function is suitable for large  datasets (e.g., UK Biobank) in plink format.
#'  Note that our code to process plink files builds from the
#' \code{\link{genio}} R package
#'
#' @param y matrix of traits (n observations by k traits)
#' @param file path to plink files
#' @param adjustment matrix of covariates to adjust traits
#' @param pop_struct matrix of PCs that captures population structure
#' @param verbose If TRUE (default) print progress.
#'
#' @return
#' A data frame of p-values where the columns are
#'  \itemize{
#'  \item{\code{wlit}: }{LIT using a linear kernel}
#'  \item{\code{ulit}: }{LIT using a projection kernel}
#'  \item{\code{alit}: }{Cauchy combination test of the above two LIT implementations.}
#' }
#'
#' @examples
#' # set seed
#' set.seed(123)
#'
#' # path to plink files
#' file <- system.file("extdata", 'sample.bed', package = "genio", mustWork = TRUE)
#'
#' # Generate trait expression
#' Y <- matrix(rnorm(10*4), ncol = 4)
#'
#' out <- lit_plink(Y, file = file)
#'
#' @seealso \code{\link{lit}}
#' @export
lit_plink <- function(y,
                    file,
                    adjustment = NULL,
                    pop_struct = NULL,
                    verbose = TRUE) {
  ## Made changes to genio::read_bed for our method
  if (missing(file))
    stop('Output file path is required!')
  file <- sub("\\.bed$", "", file)
  require_files_plink(file)
  file_bim = paste0(file, ".bim")
  file_fam = paste0(file, ".fam")
  file_bed = paste0(file, ".bed")

  # Get number of snps/ind
  bim <- genio::read_bim(file_bim, verbose = verbose)
  fam <- genio::read_fam(file_fam, verbose = verbose)
  n_ind <- nrow(fam);
  m_loci <- nrow(bim);

  if (verbose)
    message('Reading: ', file_bed)

  file <- path.expand(file_bed)
  if ( !file.exists( file_bed ) )
    stop( 'File does not exist: ', file_bed )


  if (!(is.matrix(y))) {
    stop("y input must be matrix.")
  }
  if (!(nrow(y) == n_ind)) {
    stop("Observations are rows for y.")
  }
  if (anyNA(y)) {
    stop("Current implementation requires no NAs in trait matrix. Need to adjust inputs accordingly ... feature will be added soon.")
  }

  if (is.null(pop_struct)) {
    h = matrix(1, nrow = nrow(y), ncol = 1)
  } else {
    if (!(is.matrix(pop_struct) & (nrow(y) == nrow(pop_struct)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
    h = cbind(1, pop_struct)
  }

  if (is.null(adjustment)) {
     y <- .quick_lm_cpp(Xs = h, Ys = y)
    } else {
    if (!(is.matrix(adjustment) & (nrow(y) == nrow(adjustment)))){
      stop("pop_struct must be matrix and rows are observations.")
    }
    y <- .quick_lm_cpp(Xs = cbind(h, adjustment), Ys = y)
  }
  lit_out <- .lit_bed_cpp(file = file_bed,
                          m_loci = m_loci,
                          n_ind = n_ind,
                          sY = y,
                          sH = h,
                          verbose = verbose)
  P <- lit_out$P
  colnames(P) <- c("wlit", "ulit", "alit")
  P <- as.data.frame(P);
  bim <- bim[,-3] # drop posg
  bim$maf <- lit_out$MAF
  id <- bim$maf > 0.5
  bim$maf[id] <- 1 - bim$maf[id]
  out <- cbind(bim, P)
  return(out)
}

#' @import stats
#' @import genio
#' @import CompQuadForm
#' @importFrom Rcpp sourceCpp
#' @useDynLib lit
NULL
